lms/views/api/canvas/authorize.py

```diff
@@ -1,52 +1,75 @@
-                    "state": CanvasOAuthCallbackSchema(request).state_param(),
-                }
-            ),
-            "",
+class CanvasAPIAuthorizeViews:
+    def __init__(self, request):
+        self.request = request
+        self.ai_getter = request.find_service(name="ai_getter")
+        self.canvas_api_client = request.find_service(name="canvas_api_client")
+
+    @view_config(permission="canvas_api", route_name="canvas_api.authorize")
+    def authorize(self):
+        consumer_key = self.request.lti_user.oauth_consumer_key
+
+        authorize_url = urlunparse(
+            (
+                "https",
+                urlparse(self.ai_getter.lms_url(consumer_key)).netloc,
+                "login/oauth2/auth",
+                "",
+                urlencode(
+                    {
+                        "client_id": self.ai_getter.developer_key(consumer_key),
+                        "response_type": "code",
+                        "redirect_uri": self.request.route_url("canvas_oauth_callback"),
+                        "state": CanvasOAuthCallbackSchema(self.request).state_param(),
+                    }
+                ),
+                "",
+            )
         )
-    )
 
-    return HTTPFound(location=authorize_url)
+        return HTTPFound(location=authorize_url)
+
+    @view_config(
+        permission="canvas_api",
+        renderer="lms:templates/api/canvas/oauth2_redirect.html.jinja2",
+        schema=CanvasOAuthCallbackSchema,
+    )
+    def oauth2_redirect(self):
+        authorization_code = self.request.parsed_params["code"]
 
+        try:
+            token = self.canvas_api_client.get_token(authorization_code)
+        except CanvasAPIServerError as err:
+            raise HTTPInternalServerError(
+                "Authorizing with the Canvas API failed"
+            ) from err
 
-@view_config(
-    feature_flag="new_oauth",
-    request_method="GET",
-    route_name="canvas_oauth_callback",
-    renderer="lms:templates/api/canvas/oauth2_redirect.html.jinja2",
-    schema=CanvasOAuthCallbackSchema,
-)
-def oauth2_redirect(request):
-    canvas_api_client = request.find_service(name="canvas_api_client")
+        self.canvas_api_client.save_token(*token)
 
-    authorization_code = request.parsed_params["code"]
+        return {}
 
-    token = canvas_api_client.get_token(authorization_code)
-    canvas_api_client.save_token(*token)
+    @exception_view_config(
+        renderer="lms:templates/api/canvas/oauth2_redirect_error.html.jinja2"
+    )
+    def oauth2_redirect_error(self):
+        self.request.response.status_code = 500
 
-    return {}
+        authorization_param = (
+            BearerTokenSchema(self.request).authorization_param(self.request.lti_user),
+        )
+        return {
+            "authorize_url": self.request.route_url(
+                "canvas_api.authorize", _query=[("authorization", authorization_param)]
+            )
+        }
```

Comment:
Usually exception views go in `lms/views/error.py` or `lms/views/api/error.py` because they're usually generic, but in this case we have a specific error view for the popup window (an HTML page with an error message and <kbd>Cancel>/kbd> and <kbd>Try again </kbd> buttons with specific actions). It's literally an error view for just one view, and shares view defaults with that view. So I've put it in the class with the view that it belongs to